lib2/provider/awssamlprovider.go

```diff
@@ -0,0 +1,450 @@
+// manages auth sessions for Okta applications
+package provider
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/segmentio/aws-okta/internal/sessioncache"
+	"github.com/segmentio/aws-okta/lib"
+
+	"github.com/99designs/keyring"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	aws_session "github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/sts"
+	log "github.com/sirupsen/logrus"
+	// use xerrors until 1.13 is stable/oldest supported version
+	"golang.org/x/xerrors"
+)
+
+const (
+	MaxSessionDuration    = time.Hour * 24 * 90
+	MinSessionDuration    = time.Minute * 15
+	MinAssumeRoleDuration = time.Minute * 15
+	MaxAssumeRoleDuration = time.Hour * 12
+
+	DefaultSessionDuration    = time.Hour * 4
+	DefaultAssumeRoleDuration = time.Minute * 15
+)
+
+type SessionCacheInterface interface {
+	Get(sessioncache.Key) (*sessioncache.Session, error)
+	Put(sessioncache.Key, *sessioncache.Session) error
+}
+
+type OktaClient interface {
+	AuthenticateUser() error
+	GetSessionToken() string
+	SaveSessionCookie() error
+	Request(method string, path string, queryParams url.Values, data []byte, format string, followRedirects bool) (*http.Response, error)
+	GetURL(string) (*url.URL, error)
+}
+
+type AwsSamlProvider struct {
```

Comment:
Let's take this opportunity to rename this `AWSSAMLProvider`

Reply:
And anywhere else where we have initialisms. https://github.com/golang/go/wiki/CodeReviewComments#initialisms